### PR TITLE
feat(python,rust): Dedicated horizontal aggregation functions

### PR DESCRIPTION
**Pull request by stinodego**
Addresses #8646

Changes:
* Add the following expression functions:
  * `all_horizontal`
  * `any_horizontal`
  * `max_horizontal`
  * `min_horizontal`
  * `sum_horizontal`
  * `cumsum_horizontal`
* Deprecate this horizontal functionality from the existing overloaded aggregation functions `all`, `any`, etc.
* Deprecate the functionality for passing `Series` into `all`, `any`, etc. Users should simply use `Series.all()` and similar.
* Align the Rust side

Following this will be a PR that updates the original functions `all`, `any` etc. which will close the linked issue.

### CODE REVIEW

**Comment by ritchie46** [comment]
Nice, though can we make the default `all`, `any` vertical without deprecating them. Vertical is the default in polars, so that doesn't need explicitness.

**Comment by stinodego** [comment]
> Nice, though can we make the default `all`, `any` vertical without deprecating them. Vertical is the default in polars, so that doesn't need explicitness.

Yes that's what I did! Only the part that does horizontal computation will trigger the deprecation warning.

**Review by MarcoGorelli — COMMENTED** [review]
Truly wonderful, great improvement in my opinion! 🙌 

My only concern is the low (57%) test coverage of `vertical.py` - for example, it seems there's now no tests for vertical `pl.cumsum`?

```
polars/functions/aggregation/vertical.py         67     25     40     11    57%   89-104, 160-165, 169-174, 263-268, 359-364, 456-461, 519-535
```

perhaps a little test which parametrises over these functions and the different cases would help increase confidence in the refactoring part?

**Comment by avimallu** [comment]
Will it be a good idea to also add in a blanket set of options to ignore or propagate null values? This will help situations where null is ignored for some operations (eg. sum) but not for others (eg. min/max). Reference https://github.com/pola-rs/polars/issues/9642.

**Comment by stinodego** [comment]
Yeah, we can add some convenience flags/options. But not as part of this PR :)

**Review by ritchie46 — APPROVED** [review]
Thanks @stinodego. This looks good and will save a lot of footguns in the future.

**Review by MarcoGorelli — APPROVED** [review]
Awesome! I was just preparing some teaching material and was thinking "`pl.cumsum` is going to be hard to teach clearly", and then it turned out you'd already addressed that 🥳

**Comment by ritchie46** [comment]
@stinodego 

Can we undeprecate the `pl.all()` selector?

**Comment by MarcoGorelli** [comment]
Just out of interest, why?

**Comment by ritchie46** [comment]
> Just out of interest, why?

Is that on my question? ^^

I was under the impression we want to deprecate `pl.all(**expression)` as a horizontal `Fn(bool) -> bool`. 
I do want the `all()` as an alias for `col("*")`.

**Comment by stinodego** [comment]
It shouldn't be deprecated. Let me check.

File: py-polars/polars/functions/aggregation/vertical.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+    │ ---   │
+    │ bool  │
+    ╞═══════╡
+    │ false │
+    └───────┘
+
+    """  # noqa: W505
+    if not more_exprs:
+        if exprs is None:
+            return F.col("*")
```
Comment (stinodego) [inline py-polars/polars/functions/aggregation/vertical.py:88]: Calling `all()` without any inputs should trigger this code path and not throw any DeprecationWarnings.

Comment (ritchie46) [reply py-polars/polars/functions/aggregation/vertical.py:88]: Hmm.. My IDE (jetbrains) claims I call a deprecated function. It indeed doesn't get a warning. Maybe the deprecated alias is too broad?

**Pull request merged**